### PR TITLE
Text

### PR DESCRIPTION
**Pull request by rossberg**
Fix a couple of parser omissions regarding elem & data syntax. Refactor grammar in spec slightly for more clarity. Adjust tests.

Baseline is #115. See [last commit](https://github.com/WebAssembly/bulk-memory-operations/commit/5a18820fff219cea056b827dcb6e0cc61d7f59a1) for new changes.

### CODE REVIEW

File: test/core/elem.wast [hunk 1]
```diff
@@ -5,17 +5,29 @@
...
+
+  ;; Active
+  (elem (table $t) (i32.const 0) funcref)
+  (elem (table $t) (i32.const 0) funcref (ref.func $f) (ref.null))
+  (elem (table $t) (i32.const 0) func)
+  (elem (table $t) (i32.const 0) func $f $g)
+  (elem (table $t) (i32.const 0) funcref)
+  (elem (table $t) (i32.const 0) funcref (ref.func $f) (ref.null))
+  (elem (table $t) (i32.const 0) func)
+  (elem (table $t) (i32.const 0) func $f $g)
```
Comment (gahaas) [inline test/core/elem.wast]: Aren't these just duplicates of the tests above?

File: test/core/data.wast [hunk 2]
```diff
@@ -8,14 +8,26 @@
...
+  (data $d3 (offset (i32.const 0)))
+  (data $d4 (offset (i32.const 0)) "" "a" "bc" "")
+  (data $d5 (memory 0) (i32.const 0))
+  (data $d6 (memory 0x0) (i32.const 1) "a" "" "bcd")
+  (data $d7 (memory 0x000) (offset (i32.const 0)))
+  (data $d8 (memory 0) (offset (i32.const 0)) "" "a" "bc" "")
+  (data $d9 (memory $m) (i32.const 0))
+  (data $d10 (memory $m) (i32.const 1) "a" "" "bcd")
+  (data $d11 (memory $m) (offset (i32.const 0)))
+  (data $d12 (memory $m) (offset (i32.const 0)) "" "a" "bc" "")
```
Comment (gahaas) [inline test/core/data.wast]: These tests look exactly the same as above, although with IDs. Do you expect these additional data segments to increase coverage? Could you add a comment which describes the goals of these tests?

File: test/core/elem.wast [hunk 3]
```diff
@@ -24,15 +36,41 @@
...
+  (elem $a17 (table $t) (offset (i32.const 0)) func)
+  (elem $a18 (table $t) (offset (i32.const 0)) func $f $f)
+  (elem $a19 (offset (i32.const 0)))
+  (elem $a20 (offset (i32.const 0)) funcref (ref.func $f) (ref.null))
+  (elem $a21 (offset (i32.const 0)) func $f $f)
+  (elem $a22 (offset (i32.const 0)) $f $f)
+  (elem $a23 (i32.const 0))
+  (elem $a24 (i32.const 0) funcref (ref.func $f) (ref.null))
+  (elem $a25 (i32.const 0) func $f $f)
+  (elem $a26 (i32.const 0) $f $f)
```
Comment (gahaas) [inline test/core/elem.wast]: Also here you add tests for active element segments with IDs which are the same as the tests above. What is the purpose of these tests? As far as I understand, you cannot even use active elements by ID. Can you add a comment which explains the idea behind these tests, and how they are supposed to increase coverage?

File: test/core/data.wast (hunk 2 above)
Comment (rossberg) [reply test/core/data.wast]: Yes. Note that the header comment above says "Syntax", so this module is all about testing that the parser understands all combinations of syntax.

File: test/core/elem.wast (hunk 1 above)
Comment (rossberg) [reply test/core/elem.wast]: Oops, yes, removed (here and below).

File: test/core/elem.wast (hunk 3 above)
Comment (rossberg) [reply test/core/elem.wast]: Like above, this is purely a syntax test, see heading.

Active elements can be used by ids, but behave like a dropped segment, i.e., trap in most cases.

**Pull request merged**